metricflow/dataflow/optimizer/source_scan/cm_branch_combiner.py

```diff
@@ -0,0 +1,437 @@
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        return self._default_handler(node)
+
+    def visit_aggregate_measures_node(  # noqa: D
+        self, node: AggregateMeasuresNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        current_right_node = node
+
+        combined_parent_nodes = self._combine_parent_branches(current_right_node)
+        if combined_parent_nodes is None:
+            return ComputeMetricsBranchCombinerResult()
+
+        if not isinstance(self._current_left_node, current_right_node.__class__):
+            self._log_combine_failure(
+                left_node=self._current_left_node,
+                right_node=current_right_node,
+                combine_failure_reason=f"left node is not a {current_right_node.__class__.__name__}",
+            )
+            return ComputeMetricsBranchCombinerResult()
+
+        assert len(combined_parent_nodes) == 1
+        combined_parent_node = combined_parent_nodes[0]
+        assert combined_parent_node is not None
+
+        combined_metric_input_measure_specs = (
+            self._current_left_node.metric_input_measure_specs + current_right_node.metric_input_measure_specs
+        )
+
+        for spec in combined_metric_input_measure_specs:
+            # Avoid combining branches if the AggregateMeasuresNode specifies a metric with an alias to avoid
+            # collisions e.g. two metrics use the same alias for two different measures. This is not always the case,
+            # so this could be improved later.
+            if spec.alias is not None:
+                self._log_combine_failure(
+                    left_node=self._current_left_node,
+                    right_node=current_right_node,
+                    combine_failure_reason=f"Metric input measure spec {spec} has an alias",
+                )
+                return ComputeMetricsBranchCombinerResult()
+
+        combined_node = AggregateMeasuresNode[SourceDataSetT](
+            parent_node=combined_parent_node,
+            metric_input_measure_specs=combined_metric_input_measure_specs,
+        )
+        self._log_combine_success(
+            left_node=self._current_left_node,
+            right_node=current_right_node,
+            combined_node=combined_node,
+        )
+        return ComputeMetricsBranchCombinerResult(combined_node)
+
+    def visit_compute_metrics_node(  # noqa: D
+        self, node: ComputeMetricsNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        current_right_node = node
+        self._log_visit_node_type(current_right_node)
+        combined_parent_nodes = self._combine_parent_branches(current_right_node)
+        if combined_parent_nodes is None:
+            return ComputeMetricsBranchCombinerResult()
+
+        if not isinstance(self._current_left_node, current_right_node.__class__):
```

Comment:
we could probably have a helper for this, it seems like we would do this alot here. There's already 3 instances of the exact same code here.